### PR TITLE
Pull request for #769

### LINKED ISSUE #769: Port the latest web3j release (on master) to Java 1.6 in preparation for an Android release
# Background
Provide a port of the web3j 4.0 code to Java 1.6 in order to support an Android release.
The last web3j release for android was version 3.3.1 (refer to the "master-android" branch), a new release for Android needs to be cut.

# Objective
To port the Java 1.7 and 1.8 portions of the web3j codebase to 1.6 equivalent constructs

# Requirements
- Port the code on "master" to be Java 1.6 compatible
- Ensure all existing tests run without error

# Technical details
- The last android release was version 3.3.1 and was cut from the "master-android" branch
- The current production release is 4.0.0, and is available on the "master" branch
- An attempt has been made to do this port, this can be found on "release/4.0-android"
- Most of the remaining work revolves around the usage of Java 1.8's CompletableFuture

# Next steps
There are two options.
1. Option 1: Check out "master" and port all Java 1.7 and 1.8 code to 1.6 yourself
2. Option 2: Check out "release/4.0-android", which contains a (compiling) attempt to port master to Java 1.6, and port the remaining classes

Should you choose the second path ...
- release/4.0-android contains code with the straighforward conversions performed, such as streams, optionals, etc
- we have provided a stubbed CompletableFuture at: org.web3j.backport.concurrent.CompletableFuture; you need to refactor all uses of this stub class and provide Java 1.6-compatable equivalents (such as Futures or other)
- after you address the few classes that use CompletableFuture you can delete this implementation
- you can refactor the method signatures/return types (where needed)

### PR DESCRIPTION
**Pull request by serso**
What does this PR do?
Makes web3j compatible with Android.
The project is built against the reduced JDK to ensure that it works fine on Android.
Fixes #769.

Where should the reviewer start?
First commit. I tried to make commits atomic and simple.

Why is it needed?
To make web3j available on Android.

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -6,6 +6,9 @@ dependencies {
     compile project(':abi'),
             project(':crypto'),
             project(':tuples'),
+            project(':compat'),
+            "net.sourceforge.streamsupport:streamsupport:1.7.0",
+            "net.sourceforge.streamsupport:streamsupport-cfuture:1.7.0",
```
Comment (snazha-blkio) [inline core/build.gradle]: @serso , can you please externalise the version number into a property?

**Comment by snazha-blkio** [comment]
I am happy with the approach taken here (just one minor change requested).
@antonydenyer , did you have any thoughts re: streamsupport library use and its licensing implications?
@iikirilov , can you please review and approve if you are happy?

**Comment by serso** [comment]
@snazha-blkio https://github.com/serso/web3a was created solely to test changes in web3j. I thought it would be too much to make it a Gradle submodule of web3j as then everyone gets it on checkout. It's possible to make it a git submodule though and then checkout on demand but that would require some more work.
Alternatively, it can be just forked/moved to web3j repository. I don't mind that.

@snazha-blkio do you think it's okay to merge this PR on master and force other developers to write Android-compatible Java code instead of porting every release to Android?

**Comment by serso** [comment]
@iikirilov I suppose I should rebase this branch then. What base should I use: `release/4.0` or the new `release/4.1`?

**Comment by iikirilov** [comment]
up to @snazha-blkio

File: core/build.gradle (hunk 1 above)
Comment (serso) [reply core/build.gradle]: Fixed here 10487e6b94578cb62317a141769df95037bdf452

**Comment by serso** [comment]
Okay, rebased on `release/4.1` for now.

**Comment by snazha-blkio** [comment]
Hi @serso 

Thanks for this PR, it looks great!
Could you please rebase from release/4.1-android? This is really just a clone of release/4.1, but specifically for releasing android. The changes are minimal between 4.0 and 4.1 (so far), so if you can make this change we can get a build out soon.

Thanks again!

**Comment by serso** [comment]
@snazha-blkio np, thank you for the great library.

I saw that you have already rebased the branch. I ran `./gradlew check` locally on `release/4.1-android` and it seemed to work. Am I done here? Do you need help with testing it with https://github.com/serso/web3a?

**Comment by snazha-blkio** [comment]
> @snazha-blkio np, thank you for the great library.
> 
> I saw that you have already rebased the branch. I ran `./gradlew check` locally on `release/4.1-android` and it seemed to work. Am I done here? Do you need help with testing it with https://github.com/serso/web3a?

Thanks @serso. I'll be publishing a snapshot shortly, and I will ping you when I'm ready to do some preliminary testing on a device. I'll be making use of your web3a project. 👍

**Comment by serso** [comment]
Okay, thanks. Ping me if you have any further questions regarding web3j and Android integration.

**Comment by snazha-blkio** [comment]
4.1.0-android-SNAPSHOT has been published into https://oss.sonatype.org/content/repositories/snapshots/org/web3j

**Comment by conor10** [comment]
This pull request has been mentioned on **Web3 Labs Community**. There might be relevant details there:

https://community.web3labs.com/t/web3j-android-branch-harmonisation-with-master/24/2

**Pull request merged**